modules/geo-layers/src/mvt-layer/mvt-layer.js

```diff
@@ -84,6 +80,110 @@ export default class MVTLayer extends TileLayer {
       feature => getFeatureUniqueId(feature, uniqueIdProperty) === featureIdToHighlight
     );
   }
+
+  updateState({props, oldProps, context, changeFlags}) {
+    super.updateState({props, oldProps, context, changeFlags});
+    this._debouncedGetViewportFeatures();
+  }
+
+  _getModelMatrixScale(tile) {
+    const worldScale = Math.pow(2, tile.z);
+    const xScale = WORLD_SIZE / worldScale;
+    const yScale = -xScale;
+
+    return [xScale, yScale, 1];
+  }
+
+  _getCoordinateOrigin(tile) {
+    const worldScale = Math.pow(2, tile.z);
+    const xOffset = (WORLD_SIZE * tile.x) / worldScale;
+    const yOffset = WORLD_SIZE * (1 - tile.y / worldScale);
+
+    return [xOffset, yOffset, 0];
+  }
+
+  _debouncedGetViewportFeatures() {
+    const {onViewportFeatures} = this.props;
+
+    if (onViewportFeatures) {
+      let {viewportChangeTimer} = this.state;
+      clearTimeout(viewportChangeTimer);
+
+      viewportChangeTimer = setTimeout(() => {
+        const {tileset} = this.state;
+        const {isLoaded} = tileset;
+
+        if (isLoaded) {
+          this._getViewportFeatures();
+        }
+      }, VIEWPORT_CHANGE_DEBOUNCE);
+
+      this.setState({viewportChangeTimer});
+    }
+  }
+
+  async _getViewportFeatures() {
+    const {tileset} = this.state;
+    const {onViewportFeatures} = this.props;
+    const featureCache = new Set();
+    const currentFrustumPlanes = this.context.viewport.getFrustumPlanes();
+    let viewportFeatures = [];
+
+    await tileset.selectedTiles.forEach(async tile => {
+      const features = (await tile.data) || [];
+      const transformationMatrix = new Matrix4()
+        .translate(this._getCoordinateOrigin(tile))
+        .scale(this._getModelMatrixScale(tile));
+
+      viewportFeatures = viewportFeatures.concat(
```

Comment:
Nits:
- concat tends to be fairly inefficient, compares to array.push - it copies the array again and again. Lots of memory / GC pressure and usually lower perf.
- I'd just do `for (const tile of ...) { for (const feature of ... ) if (...) features.push() }}`
- I'd allow the app to provide a `props.maxFeatures`, and stop concatenating after that number (depends on use case, but what is the point of showing 100K item lists in the UI?).
- Alternatively, it could be interesting to add a benchmark test.

Reply:
Thank you @ibgreen,

I didn't know about Deck.pickObjects. I've done a quick test and apparently it works fine with tiled layers so of course is a better solution than using math

> I'd allow the app to provide a props.maxFeatures, and stop concatenating after that number (depends on use case, but what is the point of showing 100K item lists in the UI?).

When a user wants to know the viewport features it's not always to show them in the UI. For example, you might want to display a histogram based on a property, calculate an average, max or min value, etc. I think it might be a good idea to return all features and have the user decide what to do with them.

Before modifying this PR I think it would be good to decide the design of the API, @Pessimistress talk about this in the RFC

Reply:
You could still add add a maxFeatures prop and support a (default) value of null or 0 meaning no limit

Reply:
You could still add add a maxFeatures prop and support a (default) value of null or 0 meaning no limit